cmd/healthcheck/framework/healthcheck.go

```diff
@@ -0,0 +1,445 @@
+		h, err := NewHealthCheck(options)
+		if err != nil {
+			glog.Errorf("Error initialzing: %v", err)
+			os.Exit(1)
+		}
+
+		// Start the HTTP server that enables us to serve /healtz and /metrics.   The  metrics can be pulled,
+		// analyzed and alerted on.
+		err = ServeHTTP(options)
+		if err != nil {
+			glog.Errorf("Error starting HTTP: %v", err)
+			os.Exit(1)
+		}
+
+		glog.Infof("Scheduled health checks will be run every %v", options.HealthCheckInterval)
+
+		// Every X interval run the health check
+		ticker := time.NewTicker(options.HealthCheckInterval)
+		for range ticker.C {
+			h.RunHealthCheck(options)
+		}
+	},
+}
+
+// HealthCheck is a type that used to control various aspects of the health
+// check.
+type HealthCheck struct {
+	kubeClientSet           kubernetes.Interface
+	serviceCatalogClientSet clientset.Interface
+	brokername              string
+	brokernamespace         string
+	serviceclassName        string
+	serviceclassID          string
+	serviceplanID           string
+	instanceName            string
+	bindingName             string
+	brokerendpointName      string
+	namespace               *corev1.Namespace
+	frameworkError          error
+}
+
+// NewHealthCheck creates a new HealthCheck object and initializes the kube
+// and catalog client sets.
+func NewHealthCheck(s *HealthCheckServer) (*HealthCheck, error) {
+	h := &HealthCheck{}
+	var kubeConfig *rest.Config
+
+	err := h.initBrokerAttributes(s)
+	if err != nil {
+		return nil, err
+	}
+
+	// If token exists assume we are running in a pod
+	_, err = os.Stat("/var/run/secrets/kubernetes.io/serviceaccount/token")
+	if err == nil {
+		kubeConfig, err = rest.InClusterConfig()
+	} else {
+		kubeConfig, err = LoadConfig(s.KubeConfig, s.KubeContext)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	h.kubeClientSet, err = kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		glog.Errorf("Error creating kubeClientSet: %v", err)
+		return nil, err
+	}
+
+	h.serviceCatalogClientSet, err = clientset.NewForConfig(kubeConfig)
+	if err != nil {
+		glog.Errorf("Error creating serviceCatalogClientSet: %v", err)
+		return nil, err
+	}
+
+	return h, nil
+}
+
+// RunHealthCheck runs an end to end verification against the "ups-broker".  It
+// validates the broker endpoint is available, then creates an instance and
+// binding and does validation along the way and then tears it down.  Some basic
+// Prometheus metrics are maintained that can be alerted off from.
+func (h *HealthCheck) RunHealthCheck(s *HealthCheckServer) error {
+	ExecutionCount.Inc()
+	hcStartTime := time.Now()
+
+	h.verifyBrokerIsReady()
+	h.createNamespace()
+	h.createInstance()
+	h.createBinding()
+	h.deprovision()
+	h.deleteNamespace()
+
+	if h.frameworkError == nil {
+		ReportOperationCompleted("healthcheck_completed", hcStartTime)
+		glog.V(2).Infof("Successfully ran health check in %v", time.Since(hcStartTime))
+		glog.V(4).Info("") // for readabilty/separation of test runs
+	} else {
+		h.cleanup()
```

Comment:
Would it make sense to have this in a defer so that if for any reason things go wrong, we don't accidentally leave resources around?